spring-cloud-dataflow-server/src/test/java/org/springframework/cloud/dataflow/integration/test/DataFlowIT.java

```diff
@@ -379,6 +380,72 @@ public void streamPartitioning() {
 		}
 	}
 
+	@Test
+	@Order(Integer.MIN_VALUE + 10)
+	public void streamAppCrossVersion() {
+		logger.info("stream-app-cross-version-test: DEPLOY");
+
+		int CURRENT_MANIFEST = 0;
+		int VER_LOG = 1;
+		int RANDOM = new Random().nextInt();
+
+		try (Stream stream = Stream.builder(dataFlowOperations)
+				.name("app-cross-version-test" + randomSuffix())
+				.definition("http | ver-log")
+				.create()
+				.deploy(new DeploymentPropertiesBuilder()
+						.putAll(testDeploymentProperties())
+						.put("version.ver-log", "3.0.1")
+						.build())
+		) {
+
+			assertThat(stream.getStatus()).is(
+					condition(status -> status.equals(DEPLOYING) || status.equals(PARTIAL)));
+
+			Awaitility.await().until(() -> stream.getStatus().equals(DEPLOYED));
+
+			assertThat(new SpringCloudDeployerApplicationManifestReader().read(stream.manifest(CURRENT_MANIFEST))
+					.get(VER_LOG).getSpec().getVersion()).isEqualTo("3.0.1");
+
+			runtimeApps.httpPost(stream.getName(), "http", "Test message One - " + RANDOM);
+			Awaitility.await().until(() -> stream.logs(app("ver-log")).contains("Test message One - " + RANDOM));
+
+			assertThat(stream.history().size()).isEqualTo(1L);
+
+			// UPDATE
+			logger.info("stream-app-cross-version-test: UPDATE");
+
+			stream.update(new DeploymentPropertiesBuilder().put("version.ver-log", "2.1.5.RELEASE").build());
```

Comment:
Can any of these versions point to always the `master` or `snapshot` equivalent? That way, we know we are running the latest on top of the previously GA'd code. Otherwise, these versions would become stale very soon.

Reply:
Good point, but even if we switch to `SNAPSHOT` we need to manually edit to change the SNAPSHOT version as well. In the long run, we could make it more configurable.